File: js/curve.js
// Reference implementation of the bonding curve in onchain-coins-spec.md §3.3.
//
// This is the executable version of the spec's math, written before the Rust
// program exists so the rules are proven in isolation: pure integer arithmetic,
// BigInt throughout to mirror u64/u128 semantics, no blockchain anywhere. When
// the Anchor program is written it must match this module number-for-number
// (differential testing), which is a load-bearing part of the audit posture.
//
// Everything the spec decided lives here:
//   - buys and curve-origin sells are pure constant product (pump.fun-identical)
//   - the curve tracks `outstanding` so it knows curve-origin from vault-origin
//   - vault-origin proceeds are capped at MAX_TAKE_BPS of the payable pool
//   - no change is ever returned: unpayable vault overflow is retired into
//     `surplus`, outside pricing, forever
//   - the whole fee is the treasury's (D6): no split, no reserve top-up, and
//     the ceiling is the promised 1%
//   - k = (Sv + Sr) * T never decreases (ceil rounding on the token side)

export const MAX_FEE_BPS = 100n;
export const MAX_TAKE_BPS = 9_800n;
const BPS = 10_000n;

const ceilDiv = (a, b) => (a + b - 1n) / b;

export class CurveError extends Error {}

function requireCond(cond, msg) {
  if (!cond) throw new CurveError(msg);
}

// virtualSol and supply-side amounts are plain integers (lamports, token base
// units). `float` is the sellable share minted into the curve at creation;
// `outstanding` starts at zero because nobody holds curve-origin tokens yet.
export function createCurve({ virtualSol, float, feeBps = 100n }) {
  requireCond(virtualSol > 0n && float > 0n, 'bad init');
  requireCond(feeBps >= 0n && feeBps <= MAX_FEE_BPS, 'fee out of bounds');
  return {
    Sv: virtualSol,
    Sr: 0n,
    T: float,
    outstanding: 0n,
    surplus: 0n,
    feeBps,
    paidIn: 0n,
    paidOut: 0n
  };
}

export function k(c) {
  return (c.Sv + c.Sr) * c.T;
}

export function price(c) {
  return Number(c.Sv + c.Sr) / Number(c.T);
}

export function buy(c, solIn) {
  requireCond(solIn > 0n, 'zero buy');
  const kk = k(c);
  const fee = (solIn * c.feeBps) / BPS;
  const net = solIn - fee;
  const newEff = c.Sv + c.Sr + net;
  const newT = ceilDiv(kk, newEff);
  const tokensOut = c.T - newT;
  requireCond(tokensOut > 0n && tokensOut <= c.T, 'buy too small');
  c.Sr += net;
  c.T = newT;
  c.outstanding += tokensOut;
  c.paidIn += solIn;
  return { tokensOut, fee };
}

export function sell(c, tokensIn) {
  requireCond(tokensIn > 0n, 'zero sell');
  const kk = k(c);
  const eff = c.Sv + c.Sr;
  const curveOrigin = tokensIn <= c.outstanding ? tokensIn : c.outstanding;
  const vaultPart = tokensIn - curveOrigin;
  // Curve-origin sells price down to Sv at most (conservation already
  // guarantees payability); any vault-origin content raises the floor so a
  // single sell can never take more than MAX_TAKE_BPS of the payable pool.
  const floorEff = vaultPart > 0n ? c.Sv + (c.Sr - (c.Sr * MAX_TAKE_BPS) / BPS) : c.Sv;
  const pricedT = (() => {
    const wanted = c.T + tokensIn;
    if (floorEff <= 0n) return wanted;
    const cap = kk / floorEff;
    return wanted < cap ? wanted : cap;
  })();
  const retired = c.T + tokensIn - pricedT;
  let newEff = ceilDiv(kk, pricedT);
  if (newEff < floorEff) newEff = floorEff;
  const gross = eff - newEff;
  requireCond(gross > 0n, 'nothing to pay');
  const fee = (gross * c.feeBps) / BPS;
  const netOut = gross - fee;
  c.Sr -= netOut + fee;
  requireCond(c.Sr >= 0n, 'insolvent: reserve went negative');
  c.T = pricedT;
  c.outstanding -= curveOrigin;
  c.surplus += retired;
  c.paidOut += netOut;
  return { netOut, fee, retired, curveOrigin, vaultPart };
}

// The invariants every state must satisfy after every operation. The test
// suite calls this relentlessly; the Rust program asserts the same set.
export function checkInvariants(c) {
  requireCond(c.Sr >= 0n, 'negative reserve');
  requireCond(c.T > 0n, 'empty curve side');
  requireCond(c.outstanding >= 0n, 'negative outstanding');
  requireCond(c.surplus >= 0n, 'negative surplus');
  requireCond(c.paidOut <= c.paidIn, 'solvency: paid out more than was ever paid in');
  return true;
}

File: programs/streamed_coin/src/curve.rs
//! The bonding curve math, ported number-for-number from the executable
//! reference in `js/curve.js` (spec §3.3). That module is the spec;
//! this one must agree with it on every input, which `tests/differential.rs`
//! enforces by replaying generated vectors through both.
//!
//! Pure integer math, no accounts, no CPIs: `u64` state with `u128`
//! intermediates mirroring the reference's BigInt semantics. The instruction
//! handlers in `lib.rs` call these functions and then move exactly the
//! lamports and tokens the outcome says.

pub const MAX_FEE_BPS: u16 = 100;
pub const MAX_TAKE_BPS: u128 = 9_800;
const BPS: u128 = 10_000;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CurveMathError {
    BadInit,
    FeeOutOfBounds,
    ZeroBuy,
    BuyTooSmall,
    ZeroSell,
    NothingToPay,
    Insolvent,
    Overflow,
}

impl CurveMathError {
    /// The reference implementation's error strings, verbatim, so the
    /// differential vectors can assert on the exact failure and not just
    /// "it failed".
    pub fn msg(self) -> &'static str {
        use CurveMathError::*;
        match self {
            BadInit => "bad init",
            FeeOutOfBounds => "fee out of bounds",
            ZeroBuy => "zero buy",
            BuyTooSmall => "buy too small",
            ZeroSell => "zero sell",
            NothingToPay => "nothing to pay",
            Insolvent => "insolvent: reserve went negative",
            Overflow => "overflow",
        }
    }
}

type CurveResult<T> = Result<T, CurveMathError>;

/// `paid_in`/`paid_out`/`surplus` are reference bookkeeping for the solvency
/// assertions; the on-chain `Curve` account does not store them (retired
/// surplus simply sits in the token vault outside `token_reserves`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CurveState {
    pub sv: u64,
    pub sr: u64,
    pub t: u64,
    pub outstanding: u64,
    pub surplus: u64,
    pub fee_bps: u16,
    pub paid_in: u128,
    pub paid_out: u128,
}

/// The whole fee is the treasury's (D6): there is no split and no reserve
/// top-up, so `fee` is exactly what leaves for the treasury.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BuyOutcome {
    pub tokens_out: u64,
    pub fee: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SellOutcome {
    pub net_out: u64,
    pub fee: u64,
    pub retired: u64,
    pub curve_origin: u64,
    pub vault_part: u64,
}

const fn ceil_div(a: u128, b: u128) -> u128 {
    (a + b - 1) / b
}

pub fn create(virtual_sol: u64, float: u64, fee_bps: u16) -> CurveResult<CurveState> {
    if virtual_sol == 0 || float == 0 {
        return Err(CurveMathError::BadInit);
    }
    if fee_bps > MAX_FEE_BPS {
        return Err(CurveMathError::FeeOutOfBounds);
    }
    Ok(CurveState {
        sv: virtual_sol,
        sr: 0,
        t: float,
        outstanding: 0,
        surplus: 0,
        fee_bps,
        paid_in: 0,
        paid_out: 0,
    })
}

pub fn k(c: &CurveState) -> CurveResult<u128> {
    (c.sv as u128 + c.sr as u128)
        .checked_mul(c.t as u128)
        .ok_or(CurveMathError::Overflow)
}

pub fn buy(c: &mut CurveState, sol_in: u64) -> CurveResult<BuyOutcome> {
    if sol_in == 0 {
        return Err(CurveMathError::ZeroBuy);
    }
    let kk = k(c)?;
    let fee = sol_in as u128 * c.fee_bps as u128 / BPS;
    let net = sol_in as u128 - fee;
    let new_eff = c.sv as u128 + c.sr as u128 + net;
    let new_t = ceil_div(kk, new_eff);
    let tokens_out = c.t as u128 - new_t;
    if tokens_out == 0 {
        return Err(CurveMathError::BuyTooSmall);
    }
    let new_sr = u64::try_from(c.sr as u128 + net).map_err(|_| CurveMathError::Overflow)?;
    let new_outstanding = (c.outstanding as u128 + tokens_out)
        .try_into()
        .map_err(|_| CurveMathError::Overflow)?;
    c.sr = new_sr;
    c.t = new_t as u64;
    c.outstanding = new_outstanding;
    c.paid_in += sol_in as u128;
    Ok(BuyOutcome {
        tokens_out: tokens_out as u64,
        fee: fee as u64,
    })
}

pub fn sell(c: &mut CurveState, tokens_in: u64) -> CurveResult<SellOutcome> {
    if tokens_in == 0 {
        return Err(CurveMathError::ZeroSell);
    }
    let kk = k(c)?;
    let eff = c.sv as u128 + c.sr as u128;
    let curve_origin = tokens_in.min(c.outstanding);
    let vault_part = tokens_in - curve_origin;
    let floor_eff = if vault_part > 0 {
        c.sv as u128 + (c.sr as u128 - c.sr as u128 * MAX_TAKE_BPS / BPS)
    } else {
        c.sv as u128
    };
    let wanted = c.t as u128 + tokens_in as u128;
    let priced_t = if floor_eff == 0 {
        wanted
    } else {
        wanted.min(kk / floor_eff)
    };
    if priced_t == 0 {
        return Err(CurveMathError::Overflow);
    }
    let retired = wanted - priced_t;
    let mut new_eff = ceil_div(kk, priced_t);
    if new_eff < floor_eff {
        new_eff = floor_eff;
    }
    let gross = eff.checked_sub(new_eff).ok_or(CurveMathError::Overflow)?;
    if gross == 0 {
        return Err(CurveMathError::NothingToPay);
    }
    let fee = gross * c.fee_bps as u128 / BPS;
    let net_out = gross - fee;
    let new_sr = (c.sr as u128)
        .checked_sub(net_out + fee)
        .ok_or(CurveMathError::Insolvent)?;
    let new_t = u64::try_from(priced_t).map_err(|_| CurveMathError::Overflow)?;
    let new_surplus =
        u64::try_from(c.surplus as u128 + retired).map_err(|_| CurveMathError::Overflow)?;
    c.sr = new_sr as u64;
    c.t = new_t;
    c.outstanding -= curve_origin;
    c.surplus = new_surplus;
    c.paid_out += net_out;
    Ok(SellOutcome {
        net_out: net_out as u64,
        fee: fee as u64,
        retired: retired as u64,
        curve_origin,
        vault_part,
    })
}

pub fn check_invariants(c: &CurveState) -> CurveResult<()> {
    if c.t == 0 {
        return Err(CurveMathError::BadInit);
    }
    if c.paid_out > c.paid_in {
        return Err(CurveMathError::Insolvent);
    }
    Ok(())
}
